src/nvim/ops.c

```diff
@@ -2637,12 +2637,86 @@ void do_put(int regname, yankreg_T *reg, int dir, long count, int flags)
-    (void)stuff_inserted((dir == FORWARD ? (count == -1 ? 'o' : 'a') :
-                          (count == -1 ? 'O' : 'i')), count, FALSE);
-    /* Putting the text is done later, so can't really move the cursor to
-     * the next character.  Use "l" to simulate it. */
-    if ((flags & PUT_CURSEND) && gchar_cursor() != NUL)
-      stuffcharReadbuff('l');
+    bool non_linewise_vis = (VIsual_active && VIsual_mode != 'V');
+
+    // PUT_LINE has special handling below which means we use 'i' to start.
+    char command_start_char = non_linewise_vis ? 'c' :
+      (flags & PUT_LINE ? 'i' : (dir == FORWARD ? 'a' : 'i'));
+
+    // To avoid being the affect of 'autoindent' on linewise puts, we create a
+    // new line with `:put _`.
+    if (flags & PUT_LINE) {
+      do_put('_', NULL, dir, 1, PUT_LINE);
+    }
+
+    // If given a count when pasting linewise, we stuff the readbuf with the
+    // dot register 'count' times split by newlines.
+    if (flags & PUT_LINE) {
+      stuffcharReadbuff(command_start_char);
+      for (; count > 0; count--) {
+        (void)stuff_inserted(NUL, 1, count != 1);
+        if (count != 1) {
+          // In order to avoid 'autoindent' affecting us if the text from the
+          // previous insert begins with whitespace, we use C-u to remove all
+          // whitespace at the beginning of the next line.
+          // We can't just insert Ctrl_U into readbuf1, because with
+          // 'backspace' including "eol" this would go back to the previous
+          // line if 'autoindent' is not on, hence we insert a dummy space just
+          // so that it can be removed with the C-u.
+          stuffReadbuff((char_u *)"\n ");
+          stuffcharReadbuff(Ctrl_U);
+        }
+      }
+    } else {
+      (void)stuff_inserted(command_start_char, count, false);
+    }
+    // Putting the text is done later, so can't move the cursor to the next
+    // character.  Simulate it with motion commands after the insert.
+    if (flags & PUT_CURSEND) {
+      if (flags & PUT_LINE) {
+        stuffReadbuff((char_u *)"j0");
+      } else {
+        // Ensure that we don't ring the bell from attempting to move into
+        // the space after the current line.
+        // We can stuff the readbuffer with "l" if
+        // 1) 'virtualedit' is "all" or "onemore"
+        // 2) We are not at the end of the line
+        // 3) We are not  (one past the end of the line && on the last line)
+        // The last condition is to account for a visual put over a
+        // selection one past the end of the line joining the current line
+        // with the one below.
+
+        // curwin->w_cursor.col marks the byte position of the cursor in the
+        // currunt line.
+        // It increases up to a max of STRLEN(ml_get(curwin->w_cursor.lnum)).
+        // When virtualedit is on, and the cursor is past the end of the line,
+        // curwin->w_cursor.coladd is incremented instead of
+        // curwin->w_cursor.col.
```

Comment:
```
        // Avoid ringing the bell from attempting to move into the space after
        // the current line. We can stuff the readbuffer with "l" if
        // 1) 'virtualedit' is "all" or "onemore"
        // 2) We are not at the end of the line
        // 3) We are not (one past the end of the line && on the last line).
        //    This accounts for a visual put over a selection one past the end
        //    of the line joining the current line with the one below.

        // curwin->w_cursor.col marks the byte position of the cursor in the
        // currunt line. It increases to a max of
        // STRLEN(ml_get(curwin->w_cursor.lnum)). With 'virtualedit' and the
        // cursor is past the end of the line, curwin->w_cursor.coladd is
        // incremented instead of curwin->w_cursor.col.
```